File: apps/sdkwork-im-h5/packages/sdkwork-im-h5-chat/src/services/chatTimelineUtils.ts
import type { TimelineViewEntry } from "@sdkwork/im-sdk";

export interface TimelinePaginationState {
  hasMore: boolean;
  nextAfterSeq: number;
}

export function resolveLatestMessageSeq(entries: readonly TimelineViewEntry[]): number {
  return entries.reduce((max, entry) => Math.max(max, entry.messageSeq ?? 0), 0);
}

export function mergeTimelineEntries(
  existing: readonly TimelineViewEntry[],
  incoming: readonly TimelineViewEntry[],
): TimelineViewEntry[] {
  const byId = new Map<string, TimelineViewEntry>();
  for (const entry of existing) {
    byId.set(entry.messageId, entry);
  }
  for (const entry of incoming) {
    byId.set(entry.messageId, entry);
  }
  return Array.from(byId.values()).sort((left, right) => left.messageSeq - right.messageSeq);
}

export function pickTimelinePagination(response: {
  hasMore?: boolean;
  nextAfterSeq?: number | null;
}): TimelinePaginationState {
  return {
    hasMore: Boolean(response.hasMore),
    nextAfterSeq: typeof response.nextAfterSeq === "number" ? response.nextAfterSeq : 0,
  };
}

File: apps/sdkwork-im-flutter-mobile/packages/sdkwork_im_flutter_mobile_chat/lib/src/services/chat_timeline_utils.dart
import 'package:sdkwork_im_flutter_mobile_core/sdkwork_im_flutter_mobile_core.dart';

class TimelinePaginationState {
  const TimelinePaginationState({
    required this.hasMore,
    required this.nextAfterSeq,
  });

  final bool hasMore;
  final int nextAfterSeq;
}

int resolveLatestMessageSeq(List<TimelineViewEntry> entries) {
  var maxSeq = 0;
  for (final entry in entries) {
    if (entry.messageSeq > maxSeq) {
      maxSeq = entry.messageSeq;
    }
  }
  return maxSeq;
}

List<TimelineViewEntry> mergeTimelineEntries(
  List<TimelineViewEntry> existing,
  List<TimelineViewEntry> incoming,
) {
  final byId = <String, TimelineViewEntry>{};
  for (final entry in existing) {
    byId[entry.messageId] = entry;
  }
  for (final entry in incoming) {
    byId[entry.messageId] = entry;
  }
  final merged = byId.values.toList()
    ..sort((left, right) => left.messageSeq.compareTo(right.messageSeq));
  return merged;
}

TimelinePaginationState pickTimelinePagination(TimelineResponse? response) {
  return TimelinePaginationState(
    hasMore: response?.hasMore ?? false,
    nextAfterSeq: response?.nextAfterSeq ?? 0,
  );
}
